java-spiffe-core/src/main/java/io/spiffe/svid/x509svid/X509Svid.java

```diff
@@ -0,0 +1,169 @@
+            throw new X509SvidException("Certificate could not be parsed from cert bytes", e);
+        }
+
+        PrivateKey privateKey;
+        try {
+            privateKey = CertificateUtils.generatePrivateKey(privateKeyBytes);
+        } catch (InvalidKeySpecException | InvalidKeyException | NoSuchAlgorithmException e) {
+            throw new X509SvidException("Private Key could not be parsed from key bytes", e);
+        }
+
+        SpiffeId spiffeId;
+        try {
+            spiffeId = CertificateUtils.getSpiffeId(x509Certificates.get(0));
+        } catch (CertificateException e) {
+            throw new X509SvidException(e.getMessage(), e);
+        }
+
+        validatePrivateKey(privateKey, x509Certificates);
+        validateLeafCertificate(x509Certificates.get(0));
+
+        if (x509Certificates.size() > 1) {
+            validateSigningCertificates(x509Certificates.subList(1, x509Certificates.size()));
+        }
+
+        return new X509Svid(spiffeId, x509Certificates, privateKey);
+    }
+
+    private static void validateSigningCertificates(final List<X509Certificate> certificates) throws X509SvidException {
+        for (X509Certificate cert : certificates) {
+            if (!CertificateUtils.isCA(cert)) {
+                throw new X509SvidException("Signing certificate must have CA flag set to true");
+            }
+            if (!CertificateUtils.hasKeyUsageCertSign(cert)) {
+                throw new X509SvidException("Signing certificate must have 'keyCertSign' as key usage");
+            }
+        }
+    }
+
+    private static void validateLeafCertificate(final X509Certificate leaf) throws X509SvidException {
+        if (CertificateUtils.isCA(leaf)) {
+            throw new X509SvidException("Leaf certificate must not have CA flag set to true");
+        }
+        validateKeyUsage(leaf);
+    }
+
+    private static void validateKeyUsage(final X509Certificate leaf) throws X509SvidException {
+        if (!CertificateUtils.hasKeyUsageDigitalSignature(leaf)) {
+            throw new X509SvidException("Leaf certificate must have 'digitalSignature' as key usage");
+        }
+        if (CertificateUtils.hasKeyUsageCertSign(leaf)) {
+            throw new X509SvidException("Leaf certificate must not have 'keyCertSign' as key usage");
+        }
+        if (CertificateUtils.hasKeyUsageCRLSign(leaf)) {
+            throw new X509SvidException("Leaf certificate must not have 'cRLSign' as key usage");
+        }
+    }
+
+    private static void validatePrivateKey(final PrivateKey privateKey, final List<X509Certificate> x509Certificates) throws X509SvidException {
+        try {
+            CertificateUtils.validatePrivateKey(privateKey, x509Certificates.get(0));
+        } catch (InvalidKeyException e) {
+            throw new X509SvidException("Private Key does not match Certificate Public Key", e);
+        }
+    }
+
+    @Override
+    public X509Svid getX509Svid() {
```

Comment:
This method doesn't make any sense to me. Why would you call this if you already have a reference to this object? i.e.
```java
X509Svid svid = X509Svid.parse(certBytes, privateKeyBytes);
// svid == svid.getX509Svid() here
```

Reply:
that method implements the interface `X509SvidSource`, so a `X509Svid` could be itself a `X509SvidSource`, but maybe it doesn't make much sense, or it is confusing at least, I will remove it.